internal/operands/tekton-cleanup/reconcile.go

```diff
@@ -0,0 +1,195 @@
+		cleanupResource[rbac.ClusterRoleList, rbac.ClusterRole],
+		cleanupResource[rbac.RoleBindingList, rbac.RoleBinding],
+		cleanupResource[v1.ServiceAccountList, v1.ServiceAccount],
+		cleanupResource[v1.ConfigMapList, v1.ConfigMap],
+	}
+	if request.CrdList.CrdExists(tektonCrd) {
+		cleanupFuncs = append(cleanupFuncs, cleanupResource[pipeline.PipelineList, pipeline.Pipeline]) //nolint:staticcheck
+		cleanupFuncs = append(cleanupFuncs, cleanupResource[pipeline.TaskList, pipeline.Task])         //nolint:staticcheck
+	}
+
+	var allResults []common.CleanupResult
+	for _, cleanupFunc := range cleanupFuncs {
+		cleanupResults, err := cleanupFunc(request)
+		if err != nil {
+			return nil, err
+		}
+		allResults = append(allResults, cleanupResults...)
+	}
+
+	return allResults, nil
+}
+
+func deprecateResource[L any, T any, PtrL interface {
+	*L
+	client.ObjectList
+}, PtrT interface {
+	*T
+	client.Object
+}](request *common.Request) error {
+	resources, err := common.ListOwnedResources[L, T, PtrL, PtrT](request, matchingLabelsOption(request.Instance))
+	if err != nil {
+		return fmt.Errorf("failed to list owned resources: %w", err)
+	}
+	for i := range resources {
+		resource := PtrT(&resources[i])
+		existingVal := resource.GetAnnotations()[tektonDeprecated]
+		if existingVal == "true" {
+			continue
+		}
+
+		resource.GetAnnotations()[tektonDeprecated] = "true"
+		if err := request.Client.Update(request.Context, resource); err != nil {
+			return fmt.Errorf("failed to update %s: %w", resource.GetObjectKind().GroupVersionKind().Kind, err)
+		}
+	}
+	return nil
+}
+
+func cleanupResource[L any, T any, PtrL interface {
+	*L
+	client.ObjectList
+}, PtrT interface {
+	*T
+	client.Object
+}](request *common.Request) ([]common.CleanupResult, error) {
+	resources, err := common.ListOwnedResources[L, T, PtrL, PtrT](request, matchingLabelsOption(request.Instance))
+	if err != nil {
+		return nil, fmt.Errorf("failed to list owned resources: %w", err)
+	}
+
+	results := make([]common.CleanupResult, 0, len(resources))
+	for i := range resources {
+		resource := PtrT(&resources[i])
+		cleanupResult, err := common.Cleanup(request, resource)
+		if err != nil {
+			return nil, fmt.Errorf("failed to cleanup resource: %w", err)
+		}
+		results = append(results, cleanupResult)
+	}
+	return results, nil
+}
+
+func matchingLabelsOption(ssp *ssp.SSP) client.MatchingLabelsSelector {
+	selector := labels.NewSelector().Add(
+		newLabelRequirementOrPanic(common.AppKubernetesManagedByLabel, selection.Equals, []string{common.AppKubernetesManagedByValue}),
+		newLabelRequirementOrPanic(common.AppKubernetesComponentLabel, selection.In, []string{common.AppComponentTektonPipelines.String(), common.AppComponentTektonTasks.String()}),
+		newLabelRequirementOrPanic(common.AppKubernetesNameLabel, selection.In, []string{operandPipelinesName, operandTasksName}),
+	)
+
+	if sspPartOf, exists := ssp.Labels[common.AppKubernetesPartOfLabel]; exists {
```

Comment:
Remind me again why this is optional, do we not add it upstream?

Reply:
HCO adds it upstream. Keeping it optional is more convenient in development environments, so we can create SSP object without it.

In my opinion requiring a label to be present on an object is not a good design. It should be in `.spec` if it contains required data.